c10/core/TensorImpl.h

```diff
@@ -138,6 +138,68 @@ struct C10_API AutogradMetaInterface {
   virtual ~AutogradMetaInterface();
 };
 
+// NOTE [ Version Counter Sharing ]
+//
+// Every Tensor has a version counter. Version counters are incremented
+// whenever the data or shape of a tensor changes through Variable operations.
+// These are typically in-place operations. Version counters are used to
+// detect modifications to saved variables which would result in incorrect
+// gradient calculations. Version counters may be shared between Variables:
+//
+// 1. A view shares the version counter of the base Variable,
+// 2. Detached variables share the version counter of the source,
+// 3. Unpacked saved variables share the version counter of the source.
+//
+// Version counters are not shared when:
+//
+// 1. We replace a `Variable`'s underlying `Tensor` by calling `set_data(...)`.
+//
+// Question 1: Why do we not increment the version counter in non-Variable
+// operations?
+//
+// Answer: We explicitly don't increment the version counter in non-Variable
```

Comment:
No, I don't think this is accurate. The most well known escape hatch is `x.data` and the fact that version counters don't get "incremented" here has nothing to do with whether or not we increment version counter in non-Variable operations; it's the fact that we didn't share the version counter in this situation.

Reply:
Also, this question comes totally out of the blue. Why would I ask a question like this? It would only make sense if I was previously told, "Hey! Version counters get updated when you call in-place operations on variables. They do NOT get updated when you call in-place operations on tensors."

Reply:
IMO: The accurate answer to this question is that we think of "version counter tracking" as an affordance that it is given to you from the Variable API. So if you don't use the Variable API, you don't get this affordance.